### PR TITLE
New geom types

### PR DESCRIPTION
**Pull request by turboladen**
Relates to #11. Adds the 2d versions of new curve/surface geometry types defined here: https://trac.osgeo.org/gdal/wiki/rfc49_curve_geometries (these were added in GDAL 2.0). Note that this doesn't add any new functionality that the new types bring; just wraps them. This also updates the inheritance hierarchy to mostly match what's given in the above link; `OGR::Geometry`--an abstract class in GDAL--used to provide class _and_ instance methods; it now just has class methods and behaves mostly like `OGRGeometryFactory` in  [this doc](https://gdal.org/api/ogrgeometry_cpp.html#). There are a handful of new mixins that are sprinkled across the geometry types too since these methods don't always follow the inheritance tree. Lastly, `OGR::Curve` and `OGR::Surface` are also allegedly abstract classes in cpp, but they're a bit odd in that the `GeoPackage` driver somehow uses them as concrete types, so they get most of the regular behavior as the other types, but just can only be instantiated using the `GeoPackage` driver.

tl;dr this adds & wraps the bare minimum of new geometry types and reorganizes the existing types to follow the cpp API.

Also:

- sets minimum Ruby to 3.0
- lots of calls that use a reference/pointer to a GDAL/OGR object now raise if they're null, but expected to be a somethign
- fixes a bunch of random things
- adds the first round of RBS typedefs

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -152,10 +191,17 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
...
+
 - [DEV-361] Move extension methods to ffi-gdal-extensions.
 - Removed attach_function to CPLURLGetValue and CPLURLAddKVP as they are not in GDAL 2.
 - `GDAL::MajorObject#description=`
 
+#### OGR
+
+- `OGR::GeometyTypes::Curve` and `OGR::GeometryTypes::Surface` removed; those used to be
+  characteristics of some geometry types, but are their own types as of GDAL 2.1.
+
```
Comment (turboladen) [inline Changelog.md:204]: Fix: I ended up adding these back as classes.

File: Steepfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  check 'lib/ogr/geometry_collection*.rb'
+  check 'lib/ogr/line_string*.rb'
+  check 'lib/ogr/linear_ring.rb'
+  check 'lib/ogr/multi_line_string*.rb'
+  check 'lib/ogr/multi_point*.rb'
+  check 'lib/ogr/multi_polygon*.rb'
+  check 'lib/ogr/none_geometry.rb'
+  check 'lib/ogr/point*.rb'
+  check 'lib/ogr/polygon*.rb'
+  check 'lib/ogr/unknown_geometry.rb'
```
Comment (turboladen) [inline Steepfile:17]: Add the other geometry types here

File: lib/gdal/internal_helpers.rb [hunk 3]
```diff
@@ -3,248 +3,266 @@
...
-      # @return [FFI::MemoryPointer]
-      def _pointer_from_data_type(data_type, size = nil)
-        pointer_type = _gdal_data_type_to_ffi(data_type)
+    # Internal factory method for returning a pointer from +variable+, which could
+    # be either of +klass+ class or a type of FFI::Pointer.
+    #
+    # @param variable [klass, FFI::MemoryPointer, FFI::Pointer]
+    # @param autorelease [Boolean] Pass this on to the pointer.
+    # @return [FFI::Pointer]
+    # @raise [FFI::GDAL::InvalidPointer]
```
Comment (turboladen) [inline lib/gdal/internal_helpers.rb:12]: Fix docs

File: lib/gdal/internal_helpers.rb [hunk 4]
```diff
@@ -3,248 +3,266 @@
...
-          FFI::Buffer.alloc_inout(pointer_type, size)
-        else
-          FFI::Buffer.alloc_inout(pointer_type)
+    # Internal factory method for returning a pointer from +variable+, which could
+    # be either of +klass+ class or a type of FFI::Pointer.  Returns +nil+ if it was
+    # unable to make the pointer.
+    #
+    # @param variable [Class, FFI::MemoryPointer, FFI::Pointer]
+    # @param autorelease [Boolean] Pass this on to the pointer.
+    # @return [FFI::Pointer, nil]
```
Comment (turboladen) [inline lib/gdal/internal_helpers.rb:30]: Fix docs

File: lib/gdal/internal_helpers.rb [hunk 5]
```diff
@@ -3,248 +3,266 @@
 module GDAL
   # @private
   module InternalHelpers
-    def self.included(base)
-      base.extend(ClassMethods)
-    end
-
-    module ClassMethods
-      # Internal factory method for returning a pointer from +variable+, which could
-      # be either of +klass+ class or a type of FFI::Pointer.
```
Comment (turboladen) [inline lib/gdal/internal_helpers.rb:12]: Lots of green/red in this file just from adjusting the indentation

File: lib/gdal/internal_helpers.rb [hunk 6]
```diff
@@ -3,248 +3,266 @@
...
+               else
+                 GDAL._read_pointer_pointer_safely(result_ptr_ptr, :string)
+               end
+
+      FFI::CPL::VSI.VSIFree(result_ptr_ptr)
+
+      result
+    end
+
+    def _cpl_read_and_free_strptr
```
Comment (turboladen) [inline lib/gdal/internal_helpers.rb:148]: docs

File: ffi-gdal.gemspec [hunk 7]
```diff
@@ -18,19 +18,19 @@ Gem::Specification.new do |spec|
   end
 
   spec.require_paths = %w[lib]
-  spec.required_ruby_version = '>= 2.6'
+  spec.required_ruby_version = '>= 3.0'
```
Comment (tindron) [inline ffi-gdal.gemspec:21]: Are we ready to do this?  Currently only kraken and twilight are on 3.0

File: .github/workflows/continuous-integration.yml [hunk 8]
```diff
@@ -28,9 +28,22 @@ jobs:
...
+    runs-on: ubuntu-18.04
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Ruby
+        uses: ruby/setup-ruby@v1
+        with:
+          bundler-cache: true
+      - name: Install dependencies
+        run: bundle install --jobs 4 --retry 3
+      - name: Rubocop
```
Comment (tindron) [inline .github/workflows/continuous-integration.yml:48]:
```suggestion
      - name: Steep
```

File: spec/unit/gdal/internal_helpers_spec.rb [hunk 9]
```diff
@@ -19,7 +15,33 @@ module Tester
...
+      let(:variable) { FFI::MemoryPointer.new(:pointer) }
+
+      it 'returns the pointer with autorelease = true' do
+        expect(variable).to receive(:autorelease=).with(true)
+        expect(subject._pointer(variable)).to eq variable
+      end
+    end
+
+    context 'variable is nil' do
+      it 'returns nil' do
```
Comment (tindron) [inline spec/unit/gdal/internal_helpers_spec.rb:32]: raises

File: lib/ogr/geometry/surface_methods.rb [hunk 10]
```diff
@@ -13,10 +15,17 @@ def area
 
       # Returns the units used by the associated OGR::SpatialReference.
       #
-      # @return [Hash]
+      # @return [{ unit_name: String, value: Float }, nil]
```
Comment (tindron) [inline lib/ogr/geometry/surface_methods.rb:18]: Is this valid yard?

File: ffi-gdal.gemspec (hunk 7 above)
Comment (turboladen) [reply ffi-gdal.gemspec:21]: IDK, seems like it's worth setting the bar here now. If kraken and twilight can go 3.0, it seems reasonable that other services shouldn't be that far off? If most services are still on 2.6 and that came out 3 years ago, I'd rather push to get us using current versions of things.

File: lib/ogr/geometry/surface_methods.rb (hunk 10 above)
Comment (turboladen) [reply lib/ogr/geometry/surface_methods.rb:18]:
```
ffi-gdal on  new-geom-types [$] via 💎 v3.0.1 took 2s
❯ yard
Files:         168
Modules:        77 (   69 undocumented)
Classes:       152 (  124 undocumented)
Constants:      86 (   71 undocumented)
Attributes:     16 (    0 undocumented)
Methods:       872 (   61 undocumented)
 72.98% documented
```
if it's not, `yard` doesn't complain about it.

Comment (turboladen) [reply lib/ogr/geometry/surface_methods.rb:18]: image
Looks alright...

File: lib/ogr/multi_line_string.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+    # Creates a polygon from a set of sparse edges.  The newly created geometry
+    # will contain a collection of reassembled Polygons.
+    #
+    # @return [OGR::Geometry] nil if the current geometry isn't a
+    #   MultiLineString or if it's impossible to reassemble due to topological
+    #   inconsistencies.
+    def polygonize
+      OGR::Geometry.build_geometry { FFI::OGR::API.OGR_G_Polygonize(@c_pointer) }
+    end
```
Comment (turboladen) [inline lib/ogr/multi_line_string.rb:25]: Explicitly return a Polygon

File: lib/ogr/multi_line_string_25d.rb [hunk 12]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+require_relative 'multi_line_string'
+require_relative 'geometry/simple_curve_25d'
+require_relative 'geometry/has_three_coordinate_dimensions'
+
+module OGR
+  # NOTE: {{#type}} will return :wkbMultiLineString (read: 2D instead of 2.5D)
+  # until a Z value is set.
```
Comment (turboladen) [inline lib/ogr/multi_line_string_25d.rb:9]: nope!

File: lib/ogr/multi_point_25d.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+require_relative 'multi_point'
+require_relative 'geometry/simple_curve_25d'
+require_relative 'geometry/has_three_coordinate_dimensions'
+
+module OGR
+  # NOTE: {{#type}} will return :wkbMultiPoint (read: 2D instead of 2.5D) until
+  # a Z value is set.
```
Comment (turboladen) [inline lib/ogr/multi_point_25d.rb:9]: byebye

File: lib/ogr/multi_polygon_25d.rb [hunk 14]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+require_relative 'multi_polygon'
+require_relative 'geometry/has_three_coordinate_dimensions'
+
+module OGR
+  # NOTE: {{#type}} will return :wkbMultiPolygon (read: 2D instead of 2.5D)
+  # until a Z value is set.
```
Comment (turboladen) [inline lib/ogr/multi_polygon_25d.rb:8]: nuke

File: lib/ogr/point_25d.rb [hunk 15]
```diff
@@ -1,18 +1,36 @@
 # frozen_string_literal: true
 
 require_relative 'point'
+require_relative 'geometry/has_three_coordinate_dimensions'
 
 module OGR
   # NOTE: {{#type}} will return :wkbPoint (read: 2D instead of 2.5D) until a Z
   # value is set.
```
Comment (turboladen) [inline lib/ogr/point_25d.rb:8]: nuke

File: lib/ogr/polygon_25d.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+require_relative 'polygon'
+require_relative 'geometry/has_three_coordinate_dimensions'
+
+module OGR
+  # NOTE: {{#type}} will return :wkbPolygon (read: 2D instead of 2.5D) until a Z
+  # value is set.
```
Comment (turboladen) [inline lib/ogr/polygon_25d.rb:8]: nuke

File: sig/interfaces.rbs [hunk 17]
```diff
@@ -0,0 +1,13 @@
+# TODO: Remove in favor of _HasAutoPointer
+interface _HasPointer
+  def c_pointer: -> FFI::Pointer
+end
+
+interface _HasAutoPointer[P]
+  def c_pointer: -> P
+end
```
Comment (turboladen) [inline sig/interfaces.rbs:8]: nuke for now.

File: sig/ogr/geometry/auto_pointer.rbs [hunk 18]
```diff
@@ -0,0 +1,7 @@
+module OGR
+  module Geometry
+    class AutoPointer < ::FFI::AutoPointer
+      def self.release: (FFI::Pointer c_pointer) -> void
+    end
+  end
+end
```
Comment (turboladen) [inline sig/ogr/geometry/auto_pointer.rbs:7]: nuke for now

File: sig/ogr/geometry/surface_methods.rbs [hunk 19]
```diff
@@ -0,0 +1,9 @@
+module OGR
+  module Geometry
+    module SurfaceMethods: _SpatialReference
+      def area: -> Float
+
+      def area_units: -> { unit_name: String, value: Float }?
+    end
+  end
+end
```
Comment (turboladen) [inline sig/ogr/geometry/surface_methods.rbs:9]: These should get moved to `OGR::Surface`

File: spec/support/shared_examples/a_geo_json_exporter.rb [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    it 'returns some String data' do
+      geo_json = subject.to_geo_json
+      expect(geo_json).to be_a String
+      expect(geo_json).to_not be_empty
+    end
+  end
+
+  describe '#to_geo_json_ex' do
+    it 'returns some String data' do
+      geo_json = subject.to_geo_json_ex
```
Comment (turboladen) [inline spec/support/shared_examples/a_geo_json_exporter.rb:14]: pass in a good & bad option?

File: spec/support/shared_examples/ogr/a_geometry.rb [hunk 21]
```diff
@@ -3,9 +3,20 @@
 RSpec.shared_examples 'a geometry' do
   require 'ogr'
 
-  describe '#coordinate_dimension' do
-    subject { geometry.coordinate_dimension }
-    it { is_expected.to eq 2 }
+  describe '#type' do
+    context 'when created with data' do
+      it 'returns :wkbMultiPoint25D' do
```
Comment (turboladen) [inline spec/support/shared_examples/ogr/a_geometry.rb:8]: fix

File: spec/support/shared_examples/ogr/a_geometry.rb [hunk 22]
```diff
@@ -3,9 +3,20 @@
...
+        expect(subject.type).to eq described_class::GEOMETRY_TYPE
+      end
+    end
+
+    context 'when created without data' do
+      subject { described_class.new }
+
+      it 'returns :wkbMultiPoint25D' do
```
Comment (turboladen) [inline spec/support/shared_examples/ogr/a_geometry.rb:16]: again

File: spec/support/shared_examples/ogr/a_gml_exporter.rb [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+    it 'returns some String data' do
+      gml = subject.to_gml
+      expect(gml).to be_a String
+      expect(gml).to_not be_empty
+    end
+  end
+
+  describe '#to_gml_ex' do
+    it 'returns some String data' do
+      gml = subject.to_gml_ex
```
Comment (turboladen) [inline spec/support/shared_examples/ogr/a_gml_exporter.rb:14]: use some options

File: spec/support/shared_examples/ogr/a_simple_curve_geometry.rb [hunk 24]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+RSpec.shared_examples 'a simple curve geometry' do
+  describe '#x' do
```
Comment (turboladen) [inline spec/support/shared_examples/ogr/a_simple_curve_geometry.rb:4]: Should be more here I think?

**Review by tindron — APPROVED** [review]

File: lib/ogr/multi_line_string.rb (hunk 11 above)
Comment (turboladen) [reply lib/ogr/multi_line_string.rb:25]: Actually, turns out this can return a polygon, multipolygon, geometrycollection and I'm not 100% sure what else. gonna leave like this for now.

File: sig/ogr/geometry/surface_methods.rbs (hunk 19 above)
Comment (turboladen) [reply sig/ogr/geometry/surface_methods.rbs:9]: Actually, no--these get mixed into MultiSurface too.

File: ffi-gdal.gemspec (hunk 7 above)
Comment (kurtronshausen) [reply ffi-gdal.gemspec:21]: I'm a little concerned that we'll be forcing services to upgrade both gdal and ruby at the same time.  Just getting gdal upgraded could be a large endeavor for logged data, even without throwing a ruby on top of it.  This also seems a little arbitrary -- why is it up to ffi-gdal to tell other services to upgrade?  Unless ffi-gdal is using ruby 3.0 specific features, we should allow it to work on 2.6.

File: Changelog.md [hunk 25]
```diff
@@ -14,17 +14,35 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
...
+  - `OGR::Surface`
+- Added `.new_from_coordinates` for `OGR::Point` and `OGR::Point25D`.
+- Added `OGR::Geometry#to_gml_ex` (well, changed `#to_gml` to wrap
+  `OGR_G_ExportToGML` instead of `OGR_G_ExportToGMLEx`.
 
 ### Changed
 
 - _BREAKING_: All methods with default boolean args are now keyword args.
-- _BREAKING_: `required_ruby_version` set to `>= 2.6`
+- _BREAKING_: `required_ruby_version` set to `>= 3.0`
```
Comment (turboladen) [inline Changelog.md:38]: revert

File: lib/ogr/multi_line_string.rb [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+      # @return [[FFI::Pointer, FFI::OGR::Core::WKBGeometryType]]
+      def polygonize(c_pointer)
+        polygon_ptr = FFI::OGR::API.OGR_G_Polygonize(c_pointer)
+
+        raise OGR::Failure, 'Error polygonizing MultiLineString' if polygon_ptr.null?
+
+        [polygon_ptr, FFI::OGR::API.OGR_G_GetGeometryType(polygon_ptr)]
+      end
+    end
+
```
Comment (turboladen) [inline lib/ogr/multi_line_string.rb:25]: don't need this anymore

File: ffi-gdal.gemspec (hunk 7 above)
Comment (turboladen) [reply ffi-gdal.gemspec]: Ok, reverted back to 2.6.
